### PR TITLE
Tidy up SM2

### PR DESCRIPTION
**Pull request by mattcaswell**
This includes the commits from #6288 (which itself included the commits from #5895), but considerably extends the work done:

- All SM2 symbols are made private and renamed to be only lowercase
- More extensive management of errors
- Remove redundant error codes
- Merge the two internal tests to remove duplicated code, and then actually call them from a recipe
- Remove values from libcrypto.num that were never made public

This PR does not attempt to extract SM2 into its own pmeth. I think that still needs to be done, but in a different PR.

### CODE REVIEW

File: crypto/sm2/sm2_crypt.c [hunk 1]
```diff
@@ -57,66 +59,92 @@ static size_t EC_field_size(const EC_GROUP *group)
...
+        SM2err(SM2_F_SM2_PLAINTEXT_SIZE, SM2_R_INVALID_DIGEST);
+        return 0;
+    }
+    if (field_size == 0) {
+        SM2err(SM2_F_SM2_PLAINTEXT_SIZE, SM2_R_INVALID_FIELD);
+        return 0;
+    }
+
+    overhead = 10 + 2 * field_size + (size_t)md_size;
+    if(msg_len <= overhead) {
```
Comment (tmshort) [inline crypto/sm2/sm2_crypt.c]: style nit: space after `if`

File: crypto/sm2/sm2_sign.c [hunk 2]
```diff
@@ -72,71 +77,82 @@ ECDSA_SIG *SM2_sig_gen(const EC_KEY *key, const BIGNUM *e)
...
-
-        BN_mod_mul(tmp, dA, r, order, ctx);
-        BN_sub(tmp, k, tmp);
-
-        BN_mod_mul(s, s, tmp, order, ctx);
+        if(!BN_add(s, dA, BN_value_one())
+                || !BN_mod_inverse(s, s, order, ctx)
+                || !BN_mod_mul(tmp, dA, r, order, ctx)
+                || !BN_sub(tmp, k, tmp)
+                || !BN_mod_mul(s, s, tmp, order, ctx) == 0) {
```
Comment (levitte) [inline crypto/sm2/sm2_sign.c]: This line doesn't seem right.  You should probably lose ` == 0`

File: crypto/sm2/sm2_crypt.c (hunk 1 above)
Comment (mattcaswell) [reply crypto/sm2/sm2_crypt.c]: Fixed

File: crypto/sm2/sm2_sign.c (hunk 2 above)
Comment (mattcaswell) [reply crypto/sm2/sm2_sign.c]: Yes - travis found the same issue. This should have caused the test to fail, but it didn't. The reason is a *different* bug in the tests! We initialised a variable "ok" to -1. If the test went through ok we set it to 1, and then returned the value of "ok". The test passed if the return value was non-zero, which of course it always was! I've fixed both issues.

**Comment by mattcaswell** [comment]
Fixup commits pushed addressing the feedback.

**Comment by mattcaswell** [comment]
@levitte - any more comments on this PR?

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**